### DISCUSSION TITLE
Loguru support

### DISCUSSION THREAD

**Discussion by noamgot** [hynek/stamina#87] [Ideas]
Hello 
I recently started using stamina, and I really like it!
I saw that there's a pre-built instrumentation support for structlog, and I wanted to know what you think about supporting [loguru](https://github.com/Delgan/loguru) alongside. 

For starters, I copied the structlog initialization code, and slightly modified it - it works:
```python
import stamina
from stamina.instrumentation import RetryDetails, RetryHook, RetryHookFactory, set_on_retry_hooks


def init_loguru() -> RetryHook:
    """
    Initialize loguru instrumentation for stamina retries.
    """
    from loguru import logger

    def log_retries(details: RetryDetails) -> None:
        log_data = dict(
            callable=details.name,
            args=tuple(repr(a) for a in details.args),
            kwargs=dict(details.kwargs.items()),
            retry_num=details.retry_num,
            caused_by=repr(details.caused_by),
            wait_for=round(details.wait_for, 2),
            waited_so_far=round(details.waited_so_far, 2),
        )
        log_data_str = " ".join(f"{key}={value}" for key, value in log_data.items())
        logger.warning(f"stamina.retry_scheduled | {log_data_str}")

    return log_retries


LoguruOnRetryHook = RetryHookFactory(init_loguru)
set_on_retry_hooks((LoguruOnRetryHook,))
i = 1


# stamina.instrumentation.set_on_retry_hooks([])
@stamina.retry(on=ValueError, attempts=3)
def func():
    global i
    print(i)
    i += 1
    raise ValueError("Oh no!")


func()

```

Output:
```
1
2024-12-30 12:49:13.688 | WARNING  | __main__:log_retries:22 - stamina.retry_scheduled | callable=__main__.func args=() kwargs={} retry_num=1 caused_by=ValueError('Oh no!') wait_for=0.65 waited_so_far=0.0
2
2024-12-30 12:49:14.340 | WARNING  | __main__:log_retries:22 - stamina.retry_scheduled | callable=__main__.func args=() kwargs={} retry_num=2 caused_by=ValueError('Oh no!') wait_for=0.38 waited_so_far=0.65
3
Traceback (most recent call last):
# ...
ValueError: Oh no!

Process finished with exit code 1
```

One open issue that I can think of is in case we have both structlog & loguru in the running environment - which one should we choose? What if the user wants to use the non-default logger?

Would love to hear your thoughts on this one...

**Comment by hynek** [comment]
Hi,

I will not maintain loguru integration. In fact, I don't plan to maintain any other implementations at all – those that are built-in are only there because I use them in all my projects. But I'm happy to help you to write/maintain yours, if you have concrete questions.

**Reply by noamgot** [reply]
Well, I wrote my question in the main post, but assuming it's not going to make it into the package, I can just use this code for myself (and leave it here in case someone will look for some implementation in the future) :)
Thanks

**Reply by hynek** [reply]
AFAICT, your question was related to auto-detection within stamina which is unless it's part of stamina proper? If I've missed another one, feel free to lmk.

**Reply by noamgot** [reply]
I suggested to:
1. integrate my code for loguru instrumentation support into stamina
2. If that happens, we need to decide whether it takes precedence over structlog or vice-versa
Hope that it makes sense, 
thanks
